routing/pathfind.go

```diff
@@ -617,6 +646,86 @@ func findPath(tx *bolt.Tx, graph *channeldb.ChannelGraph,
 	return pathEdges, nil
 }
 
+// findPathWithHints is a wrapper around findPath that attempts to find a path
+// to the target with the help of routing hints.
+//
+// NOTE: The hints must be chained together and sorted in forward order in order
+// to reach the target successfully.
+func findPathWithHints(tx *bolt.Tx, graph *channeldb.ChannelGraph,
+	sourceNode *channeldb.LightningNode, target *btcec.PublicKey,
+	hints []Hint, ignoredNodes map[Vertex]struct{},
+	ignoredEdges map[uint64]struct{},
+	amt lnwire.MilliSatoshi) ([]*ChannelHop, error) {
+
+	// If there are no hints available, we can simply find a path to the
+	// destination.
+	if len(hints) == 0 {
+		return findPath(tx, graph, sourceNode, target, ignoredNodes,
+			ignoredEdges, amt)
+	}
+
+	var finalPath []*ChannelHop
+
+	// First, we'll find a path to the starting node of the first routing
+	// hint. We should be able to do this since the first hint's starting
+	// node should either be advertised to the network, or have a direct
+	// channel with us.
+	publicTarget := hints[0].PubKey
+	path, err := findPath(tx, graph, sourceNode, publicTarget, ignoredNodes,
+		ignoredEdges, amt)
+	if err != nil {
+		return nil, err
+	}
+
+	// Add the hops from the path found above to the final path.
+	finalPath = append(finalPath, path...)
+
+	// Then, we'll go over our list of hints that should reach the
+	// target destination when chained together.
+	for i, hint := range hints {
```

Comment:
This logic seems to assume a single routing hint per invoice, and also that the routing hint spans multiple hops. Instead, due to current protocol limitations, it perhaps should instead treat each routing hint in a discrete manner. As it's possible for the receiver to have many private outgoing channels to distinct nodes. This logic would the be modified to use the first non-blacklisted-by-mission-control routing hint first, then as routes possibly fail, the rest of the routing hints would be filtered out.

Reply:
Fixed.